Add llm as judge mt-bench  dataset and metrics

OfirArviv opened the pull request.

OfirArviv commented:
@yoavkatz  @elronbandel  when this is read please go over thoroughly. There were a lot of changes, so please make sure I didn't forget anything.
These are the original Mt-Bench prompts for reference
```

{
  "name": "pair-v2",
  "type": "pairwise",
  "system_prompt": "Please act as an impartial judge and evaluate the quality of the responses provided by two AI assistants to the user question displayed below. You should choose the assistant that follows the user's instructions and answers the user's question better. Your evaluation should consider factors such as the helpfulness, relevance, accuracy, depth, creativity, and level of detail of their responses. Begin your evaluation by comparing the two responses and provide a short explanation. Avoid any position biases and ensure that the order in which the responses were presented does not influence your decision. Do not allow the length of the responses to influence your evaluation. Do not favor certain names of the assistants. Be as objective as possible. After providing your explanation, output your final verdict by strictly following this format: \"[[A]]\" if assistant A is better, \"[[B]]\" if assistant B is better, and \"[[C]]\" for a tie.",
  "prompt_template": "[User Question]\n{question}\n\n[The Start of Assistant A's Answer]\n{answer_a}\n[The End of Assistant A's Answer]\n\n[The Start of Assistant B's Answer]\n{answer_b}\n[The End of Assistant B's Answer]",
  "description": "Prompt for general questions",
  "category": "general",
  "output_format": "[[A]]"
}
{
  "name": "pair-v2-multi-turn",
  "type": "pairwise",
  "system_prompt": "Please act as an impartial judge and evaluate the quality of the responses provided by two AI assistants to the user questions. You should choose the assistant that follows the user's instructions and answers the user's questions better. Your evaluation should consider factors such as the helpfulness, relevance, accuracy, depth, creativity, and level of detail of their responses. You should focus on who provides a better answer to the second user question. Begin your evaluation by comparing the responses of the two assistants and provide a short explanation. Avoid any position biases and ensure that the order in which the responses were presented does not influence your decision. Do not allow the length of the responses to influence your evaluation. Do not favor certain names of the assistants. Be as objective as possible. After providing your explanation, output your final verdict by strictly following this format: \"[[A]]\" if assistant A is better, \"[[B]]\" if assistant B is better, and \"[[C]]\" for a tie.",
  "prompt_template": "<|The Start of Assistant A's Conversation with User|>\n\n### User:\n{question_1}\n\n### Assistant A:\n{answer_a_1}\n\n### User:\n{question_2}\n\n### Assistant A:\n{answer_a_2}\n\n<|The End of Assistant A's Conversation with User|>\n\n\n<|The Start of Assistant B's Conversation with User|>\n\n### User:\n{question_1}\n\n### Assistant B:\n{answer_b_1}\n\n### User:\n{question_2}\n\n### Assistant B:\n{answer_b_2}\n\n<|The End of Assistant B's Conversation with User|>",
  "description": "Prompt for multi-turn general questions",
  "category": "general",
  "output_format": "[[A]]"
}
{
  "name": "pair-math-v1",
  "type": "pairwise",
  "system_prompt": "Please act as an impartial judge and evaluate the quality of the responses provided by two AI assistants to the user question displayed below. Your evaluation should consider correctness and helpfulness. You will be given a reference answer, assistant A's answer, and assistant B's answer. Your job is to evaluate which assistant's answer is better. Begin your evaluation by comparing both assistants' answers with the reference answer. Identify and correct any mistakes. Avoid any position biases and ensure that the order in which the responses were presented does not influence your decision. Do not allow the length of the responses to influence your evaluation. Do not favor certain names of the assistants. Be as objective as possible. After providing your explanation, output your final verdict by strictly following this format: \"[[A]]\" if assistant A is better, \"[[B]]\" if assistant B is better, and \"[[C]]\" for a tie.",
  "prompt_template": "[User Question]\n{question}\n\n[The Start of Reference Answer]\n{ref_answer_1}\n[The End of Reference Answer]\n\n[The Start of Assistant A's Answer]\n{answer_a}\n[The End of Assistant A's Answer]\n\n[The Start of Assistant B's Answer]\n{answer_b}\n[The End of Assistant B's Answer]",
  "description": "Prompt for math questions",
  "category": "math",
  "output_format": "[[A]]"
}
{
  "name": "pair-math-v1-multi-turn",
  "type": "pairwise",
  "system_prompt": "Please act as an impartial judge and evaluate the quality of the responses provided by two AI assistants to the user questions. Your evaluation should consider correctness and helpfulness. You will be given reference answers, the assistant A's answers, the assistant B's answers. Your job is to determine which assistant provides correct and helpful answers to the second user question. Begin your evaluation by comparing both assistants' answers with the reference answers. Identify and correct any mistakes. Avoid any position biases and ensure that the order in which the responses were presented does not influence your decision. Do not allow the length of the responses to influence your evaluation. Do not favor certain names of the assistants. Be as objective as possible. After providing your explanation, output your final verdict by strictly following this format: \"[[A]]\" if assistant A is better, \"[[B]]\" if assistant B is better, and \"[[C]]\" for a tie.",
  "prompt_template": "<|The Start of Reference Answer|>\n\n### User:\n{question_1}\n\n### Reference answer:\n{ref_answer_1}\n\n### User:\n{question_2}\n\n### Reference answer:\n{ref_answer_2}\n\n<|The End of Reference Answer|>\n\n\n<|The Start of Assistant A's Conversation with User|>\n\n### User:\n{question_1}\n\n### Assistant A:\n{answer_a_1}\n\n### User:\n{question_2}\n\n### Assistant A:\n{answer_a_2}\n\n<|The End of Assistant A's Conversation with User|>\n\n\n<|The Start of Assistant B's Conversation with User|>\n\n### User:\n{question_1}\n\n### Assistant B:\n{answer_b_1}\n\n### User:\n{question_2}\n\n### Assistant B:\n{answer_b_2}\n\n<|The End of Assistant B's Conversation with User|>",
  "description": "Prompt for multi-turn general questions",
  "category": "general",
  "output_format": "[[A]]"
}
{
  "name": "single-v1",
  "type": "single",
  "system_prompt": "You are a helpful assistant.",
  "prompt_template": "[Instruction]\nPlease act as an impartial judge and evaluate the quality of the response provided by an AI assistant to the user question displayed below. Your evaluation should consider factors such as the helpfulness, relevance, accuracy, depth, creativity, and level of detail of the response. Begin your evaluation by providing a short explanation. Be as objective as possible. After providing your explanation, you must rate the response on a scale of 1 to 10 by strictly following this format: \"[[rating]]\", for example: \"Rating: [[5]]\".\n\n[Question]\n{question}\n\n[The Start of Assistant's Answer]\n{answer}\n[The End of Assistant's Answer]",
  "description": "Prompt for general questions",
  "category": "general",
  "output_format": "[[rating]]"
}
{
  "name": "single-math-v1",
  "type": "single",
  "system_prompt": "You are a helpful assistant.",
  "prompt_template": "[Instruction]\nPlease act as an impartial judge and evaluate the quality of the response provided by an AI assistant to the user question displayed below. Your evaluation should consider correctness and helpfulness. You will be given a reference answer and the assistant's answer. Begin your evaluation by comparing the assistant's answer with the reference answer. Identify and correct any mistakes. Be as objective as possible. After providing your explanation, you must rate the response on a scale of 1 to 10 by strictly following this format: \"[[rating]]\", for example: \"Rating: [[5]]\".\n\n[Question]\n{question}\n\n[The Start of Reference Answer]\n{ref_answer_1}\n[The End of Reference Answer]\n\n[The Start of Assistant's Answer]\n{answer}\n[The End of Assistant's Answer]",
  "description": "Prompt for general questions",
  "category": "math",
  "output_format": "[[rating]]"
}
{
  "name": "single-v1-multi-turn",
  "type": "single",
  "system_prompt": "Please act as an impartial judge and evaluate the quality of the response provided by an AI assistant to the user question displayed below. Your evaluation should consider factors such as the helpfulness, relevance, accuracy, depth, creativity, and level of detail of the response. You evaluation should focus on the assistant's answer to the second user question. Begin your evaluation by providing a short explanation. Be as objective as possible. After providing your explanation, you must rate the response on a scale of 1 to 10 by strictly following this format: \"[[rating]]\", for example: \"Rating: [[5]]\".\n\n",
  "prompt_template": "<|The Start of Assistant A's Conversation with User|>\n\n### User:\n{question_1}\n\n### Assistant A:\n{answer_1}\n\n### User:\n{question_2}\n\n### Assistant A:\n{answer_2}\n\n<|The End of Assistant A's Conversation with User|>",
  "description": "Prompt for general questions",
  "category": "general",
  "output_format": "[[rating]]"
}
{
  "name": "single-math-v1-multi-turn",
  "type": "single",
  "system_prompt": "Please act as an impartial judge and evaluate the quality of the response provided by an AI assistant to the user question. Your evaluation should consider correctness and helpfulness. You will be given a reference answer and the assistant's answer. You evaluation should focus on the assistant's answer to the second question. Begin your evaluation by comparing the assistant's answer with the reference answer. Identify and correct any mistakes. Be as objective as possible. After providing your explanation, you must rate the response on a scale of 1 to 10 by strictly following this format: \"[[rating]]\", for example: \"Rating: [[5]]\".\n\n",
  "prompt_template": "<|The Start of Reference Answer|>\n\n### User:\n{question_1}\n\n### Reference answer:\n{ref_answer_1}\n\n### User:\n{question_2}\n\n### Reference answer:\n{ref_answer_2}\n\n<|The End of Reference Answer|>\n\n\n<|The Start of Assistant A's Conversation with User|>\n\n### User:\n{question_1}\n\n### Assistant A:\n{answer_1}\n\n### User:\n{question_2}\n\n### Assistant A:\n{answer_2}\n\n<|The End of Assistant A's Conversation with User|>",
  "description": "Prompt for general questions",
  "category": "math",
  "output_format": "[[rating]]"
}
```

OfirArviv commented on prepare/templates/model_response_assessment/mt_bench_model_pairwise_comparison_multi_turn.py:
```diff
@@ -0,0 +1,47 @@
+        " influence your evaluation. Do not favor certain names of the assistants. Be as objective as"
+        " possible. After providing your explanation, output your final verdict by strictly"
+        ' following this format: "[[A]]" if assistant A is better, "[[B]]" if assistant B is better,'
+        ' and "[[C]]" for a tie.\n\n',
```
@elronbandel Not sure if to leave this separator to the model format

OfirArviv replied:
need to find a way to parasitize the "A", "B", and "C" options

elronbandel reviewed (changes requested):
Regarding naming, I think the right naming for the tasks are:
`evaluation.response_rating` (single turn)
`evaluation.response_rating.with_reference`
`evaluation.response_selection` or  `evaluation.response_preference`
etc

also in the task fields i would change `model_answer` to `response` 

The reason is to make sure it is general enough to systems which are not only just model. As well as having the task name fit to other tasks in the unitxt taxonomy. "evaluation" i think is the core task here, which is to evaluate responses esentially.

elronbandel commented on prepare/tasks/model_response_assessment/model_pairwise_comparison_with_reference_multi_turn.py:
```diff
@@ -0,0 +1,19 @@
+            "model_a_dialog": "List[Tuple[str, str]]",
+            "model_b_dialog": "List[Tuple[str, str]]",
+            "reference_dialog": "List[Tuple[str, str]]",
+            "model_a_label": "str",
+            "model_b_label": "str",
+            "tie_label": "str",
```
Why didnt you use them directly in the template instruction etc?

OfirArviv replied:
because we discussed you might have a better way to handle this. In the end its a multi-choice question

OfirArviv commented:
> Regarding naming, I think the right naming for the tasks are: `evaluation.response_rating` (single turn) `evaluation.response_rating.with_reference` `evaluation.response_selection` or `evaluation.response_preference` etc
> 
> also in the task fields i would change `model_answer` to `response`
> 
> The reason is to make sure it is general enough to systems which are not only just model. As well as having the task name fit to other tasks in the unitxt taxonomy. "evaluation" i think is the core task here, which is to evaluate responses esentially.

I think "evaluation"  (and the other names you proposed) are not very  descriptive. In the end the number one job of the "task name" is to be descriptive so people can easily understand what it means and to find it if thy need to.

elronbandel commented:
@OfirArviv can you use this operator: https://github.com/IBM/unitxt/pull/813#pullrequestreview-2047817508
once merged?

OfirArviv commented:
> > Regarding naming, I think the right naming for the tasks are: `evaluation.response_rating` (single turn) `evaluation.response_rating.with_reference` `evaluation.response_selection` or `evaluation.response_preference` etc
> > also in the task fields i would change `model_answer` to `response`
> > The reason is to make sure it is general enough to systems which are not only just model. As well as having the task name fit to other tasks in the unitxt taxonomy. "evaluation" i think is the core task here, which is to evaluate responses esentially.
> 
> I think "evaluation" (and the other names you proposed) are not very descriptive (evaluation of what?). In the end the number one job of the "task name" is to be descriptive so people can easily understand what it means and to find it if they need to. And even that is not ideally, as people will look for "llm_as_judge" probably. response_evaluation might be better. Also I don't think we should nest things too much (i.e. response_rating.with_reference), unless we need to because we have too many stuff in a dir. It's messy and imo not comfortable to work with or view.

We decided on: 

--Task--
response_assessment.pairwise_comparison
response_assessment.rating

pairwise_comparison
rating

--Cards--
mt_bench.pairwise_comparison
mt_bench.rating

--Template--
response_assessment.pairwise_comparison
response_assessment.rating

pairwise_comparison.model
rating.model_outputs

yoavkatz commented:
@OfirArviv - Overall loogs good to me, but tests seem to fail.

1. AssertionError: Error while trying to run IbmGenAiInferenceEngine. Please set the environment param 'GENAI_KEY'.  
We should not set an API in the CI/CD environment, but we can set a dummy key, just so the 'prepare' will pass.

024-05-12T19:32:30.4173530Z   File "/home/runner/work/unitxt/unitxt/prepare/metrics/llm_as_judge/rating/llama_3_ibm_genai_mt_bench_template.py", line 29, in <module>
2024-05-12T19:32:30.4173687Z     inference_model = IbmGenAiInferenceEngine(
2024-05-12T19:32:30.4174075Z   File "/home/runner/work/unitxt/unitxt/src/unitxt/dataclass.py", line 336, in custom_cls_init
2024-05-12T19:32:30.4174197Z     original_init(self, *args, **kwargs)
2024-05-12T19:32:30.4174573Z   File "/home/runner/work/unitxt/unitxt/src/unitxt/dataclass.py", line 336, in custom_cls_init
2024-05-12T19:32:30.4174695Z     original_init(self, *args, **kwargs)
2024-05-12T19:32:30.4175080Z   File "/home/runner/work/unitxt/unitxt/src/unitxt/dataclass.py", line 336, in custom_cls_init
2024-05-12T19:32:30.4175196Z     original_init(self, *args, **kwargs)
2024-05-12T19:32:30.4175312Z   [Previous line repeated 1 more time]
2024-05-12T19:32:30.4175722Z   File "/home/runner/work/unitxt/unitxt/src/unitxt/dataclass.py", line 502, in __init__
2024-05-12T19:32:30.4175818Z     self.__post_init__()
2024-05-12T19:32:30.4176182Z   File "/home/runner/work/unitxt/unitxt/src/unitxt/artifact.py", line 251, in __post_init__
2024-05-12T19:32:30.4176280Z     self.prepare()
2024-05-12T19:32:30.4176623Z   File "/home/runner/work/unitxt/unitxt/src/unitxt/inference.py", line 78, in prepare
2024-05-12T19:32:30.4176741Z     assert api_key is not None, (
2024-05-12T19:32:30.4177242Z AssertionError: Error while trying to run IbmGenAiInferenceEngine. Please set the environment param 'GENAI_KEY'.

OfirArviv commented:
@yoavkatz I updated the catalog test pyhon to 3.9, as bam require this as a minimum.
I kept the 3.8 for the core unitxt test so we ca still claim it support python 3.8

OfirArviv commented:
> @yoavkatz I updated the catalog test pyhon to 3.9, as bam require this as a minimum. I kept the 3.8 for the core unitxt test so we ca still claim it support python 3.8

@yoavkatz the main test still run
    - run: pip install -r requirements/tests.rqr
    which contain ibm_genai.
    But we could so is 
1) update unitxt entirely to 3.9
2) split the catalog tests from the rest of the tests

yoavkatz commented on docs/docs/llm_as_judge.rst:
```diff
@@ -85,24 +85,32 @@ of metrics.rag.model_response_assessment.llm_as_judge_by_flan_t5_large_on_hf_pip
-        "demos_pool_size=0,"
-        "num_demos=0"
+    model_id = "meta-llama/llama-3-8b-instruct"
+    gen_params = IbmGenAiInferenceEngineParams(max_new_tokens=252)
```
The example in the documentation should not be based only on IBMGenAI, because most people can not access it.

OfirArviv replied:
I'm in the middle of rewriting it
